### ISSUE TITLE
MakeRef vector does not see the same valid host/device data as the base vector

### ISSUE THREAD

**Issue by stefanhenneking** [mfem/mfem#1578]
While attempting to implement the device support for complex operators (see #1575), I ran into the issue that subvectors (real/imaginary) of a device vector did not appear to see modifications of the underlying data when the base vector was moved between host and device; this issue might be related to #1468. Below, I created a small test program that, when executed with the cuda-compiled MFEM library, demonstrates the issue: As I understand it, the expected behavior in the example would be that `Vector w` should see the same valid data as `Vector v`.

Code snippet:
```c++
   const int vSize = 10;
   Vector v; v.UseDevice(true); v.SetSize(vSize); v = 0.0;
   cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;

   Vector w; w.MakeRef(v, 0, vSize);
   cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;

   auto hv = v.HostWrite();
   for (int j = 0; j < vSize; j++) { hv[j] = 1.0; }
   cout << "IsHost(v) = " << IsHostMemory(v.GetMemory().GetMemoryType()) << endl;
   cout << "IsHost(w) = " << IsHostMemory(w.GetMemory().GetMemoryType()) << endl;

   Vector z; z.UseDevice(true); z.SetSize(vSize);
   auto dz = z.Write();
   auto dw = w.Read();
   MFEM_FORALL(i, vSize,
   {
      dz[i] = dw[i];
   });
   z.HostRead();
   cout << "norm(z) = " << z.Norml2() << endl;

   dz = z.Write();
   auto dv = v.Read();
   MFEM_FORALL(i, vSize,
   {
      dz[i] = dv[i];
   });
   z.HostRead();
   cout << "norm(z) = " << z.Norml2() << endl;
```
Output:
```
IsHost(v) = 0
IsHost(w) = 0
IsHost(v) = 1
IsHost(w) = 0
norm(z) = 0
norm(z) = 3.16228
```

**Comment by v-dobrev** [comment]
There is no easy way to keep the big "base" Vector (`v` in your example) and the "alias" sub-Vector (`w` in your example) synchronized when they are being moved/copied between host and device. Therefore such synchronizations need to be done "manually" using the methods `Vector::SyncMemory` and `Vector::SyncAliasMemory`.

Basically the issue is that the Memory objects (inside the Vectors) do not know about the other version, so they cannot update the validity flags (the host and device validity flags indicate which of the pointers has valid data) of the other Vector. Also such update may not make sense if you just moved the sub-Vector.

In your example above, after you move the "base" Vector `v` to host, you need to "inform" the "alias" `w` that the validity flags of its base have been changed. This is done by calling `w.SyncMemory(v)` which simply copies the validity flags from `v` to `w` -- there are no host-device memory transfers involved.

One the other hand, if in your example you moved `w` to host and modified it there, and then you want to access the data through the base Vector `v` (you can think of the more general case here, when `w` is smaller than `v`) then you need to call `w.SyncAliasMemory(v)`. In this particular case, the call will move the sub-Vector described by `w` from host to device and update the validity flags of `w` to be the same as the ones of `v`. The way the whole Vector `v` gets the real data in one location -- before the call part of it was on device and the part described by `w` was on host.

Both `w.SyncMemory(v)` and `w.SyncAliasMemory(v)` ensure that `w` gets the validity flags of `v`, the difference is where the real data is before the call -- in the first case the real data is in `v` and in the second, it is in `w`.

**Comment by stefanhenneking** [comment]
Thank you for the detailed explanation, @v-dobrev. I can see how there's no easy way to prevent this manual synchronization, depending on the use case. With this in mind, I should be able to place synchronization calls in the right places to assure that both the sub-vectors as well as the base can be seen as valid/synchronized from the user perspective at all times.

It's a bit of a pitfall if one is unaware of this, but I hope that other developers/users find this posting with your explanation to avoid a lengthy bug search should they run into similar issues.

**Comment by mlstowell** [comment]
@stefanhenneking , you make a good point about documentation.  If there's a place in the header files (or their Doxygen counterparts) where you think this information would have been useful we can certainly add something.  Another possibility is to add a page to [mfem.org](https://mfem.org) dedicated to the `MemoryManager` and related classes.  Such a page could describe a handful of common or tricky use cases, etc..

**Comment by stefanhenneking** [comment]
@v-dobrev One related issue that came up now is the following behavior:

```cpp
   const int vSize(10);
   Vector v; v.UseDevice(true); v.SetSize(10); v = 1.0;

   Vector w(vSize); w = 1.0;
   Vector z; z.MakeRef(w, 0, vSize);

   cout << "v.UseDevice() = " << v.UseDevice() << endl;
   cout << "w.UseDevice() = " << w.UseDevice() << endl;
   cout << "z.UseDevice() = " << z.UseDevice() << endl;
   
   cout << "#1" << endl;
   w.Add(1.0, v); // works
   cout << "#2" << endl;
   z.Add(1.0, v); // crashes
   cout << "#3" <<endl;
```
Output:
```
v.UseDevice() = 1
w.UseDevice() = 0
z.UseDevice() = 0
#1
#2

alias not found
Aborted
```
This happens, for example, when `MakeRef` is used to alias a subvector in the `Mult` operator where one of the vectors passed has `UseDevice() = false` as is the case for some solvers: for example, `GMRESSolver::Mult` calling `oper->Mult(x, r)` where `r` is a host vector with `UseDevice() = false`, and subsequently ComplexOperator has to use `MakeRef` to alias the real/imaginary part of `r`.

Is this behavior an issue that needs to be resolved, or should the operator mult that is using `MakeRef` aliases ensure that the UseDevice flag is set to the same value for both vectors?

**Comment by v-dobrev** [comment]
Yes, this is another known caveat. This error message indicates that you are trying to move an "alias" Vector to gpu while its "base" Vector did not have gpu allocation (valid or not) when the alias was created (and may still not have gpu allocation when the move of the "alias" was attempted). This is another case where we cannot update the "base" Vector because we do not have access to it, and even if we did, there are other complications.

Therefore, we need to follow this rule: if you are creating an "alias" that will be used on device, you need to ensure the "base" is allocated on device. Depending on the context, one can use different methods to do that.

For example, if the "base" is initialized (on host, otherwise there will be no issue) in the same function that will create the alias, one can call `base.Write()` to create the device allocation followed by `base.HostWrite()` and then initialize "base" on host -- this sequence avoids any unnecessary host-device transfers.

Another example: if the "base" was initialized outside of the function where the "alias" is created, then the most appropriate choice probably is to call `base.Read()` before creating the "alias". Since the alias will need the data on device, the incurred host-to-device transfer is (at least partially) necessary anyway.

Ideally, "base" Vectors that will be modified/accessed on device through aliases should be allocated on device to begin with, e.g. using `Vector::SetSize(int s, MemoryType mt)` typically with `mt = Device::GetDeviceMemoryType()`.

Just to add to @mlstowell's remark, we are in the process of gathering input for a webpage with "GPU Tips and Tricks" on mfem.org where we hope to have these types of questions and gotchas answered and summarized.

**Comment by tzanio** [comment]
@YohannDudouit, make sure to include this in your GPU FAQ page

**Comment by stefanhenneking** [comment]
@v-dobrev, @YohannDudouit I am appending a related issue that came up in a different situation but also in the context of using `MakeRef`. I was able to pinpoint the issue down to the following behavior:

```cpp
   int width = 1000000;
   Vector v_r(width/2), v_i(width/2); // this initialization causes failure below
   {   
      // Vector v_r(width/2), v_i(width/2); // this initialization is fine
      Vector v(width);
      v.UseDevice(true); 
      v = 0.0;

      v_r.MakeRef(v, 0); 
      v_i.MakeRef(v, width/2);
   }   
   Vector w(width);
   w.UseDevice(true);
   w = 0.0;
   Vector w_r(width/2), w_i(width/2);

   cout << myid << ": == # START # ==" << endl;
   MPI_Barrier(MPI_COMM_WORLD);
   w_r.MakeRef(w, 0);        // fails
   w_i.MakeRef(w, width/2);  // fails
   // with one of the following two messages:
   // 1) Verification failed: (h_mt == maps->aliases.at(h_ptr).mem->h_mt) is false:
   //    --> 
   //    ... in function: static void mfem::MemoryManager::CheckHostMemoryType_(mfem::MemoryType, void *)
   //    ... in file: general/mem_manager.cpp:1377 
   // 2) alias already exists with different base/offset! 
   MPI_Barrier(MPI_COMM_WORLD);
   cout << myid << ": == # END # ==" << endl;
```

The tricky thing here was that this happened from using `MakeRef` inside an operator `Mult` where the operator has its vectors pre-initialized on construction without data (e.g., `ComplexOperator`), and they then reference the base vectors that are passed from, say the iterative solver as an argument to `Mult`. These base vectors go out of scope when the solver finishes (similar to `v` going out of scope in the example), but the operator still exists (similar to `v_r` and `v_i` still existing when `v` goes out of scope).

Then, when allocating new vectors on device (`w` above) and referencing them (`w_r`, `w_i`), somehow these aliases (`v_r`, `v_i`) are still dangling and this can lead to these errors form the memory manager. This, at least, is how I understand what is happening.

Lastly, this error is thrown depending on various factors (e.g., other program allocations, device memory, number of MPI ranks, vector size) related to memory management. It should be relatively easy to reproduce the errors on `Ray` where I tested this in different contexts (examples, miniapps).

I'm not sure if this can be resolved from the memory manager point-of-view, but I thought I should ask about it, and, if nothing else, this might help to document it for developers.

**Comment by stefanhenneking** [comment]
After some preliminary discussions with @v-dobrev, we have summarized that this last issue can be avoided by ensuring that the memory object of the alias vector (`v_r`, `v_i` in the example) is destroyed before the base vector (`v` in the example) is deallocated. That means, either
1) the alias vector is defined within the same or a smaller scope that the base vector, or
2) the alias vector is "destroyed" before the base vector is deallocated (`v_r.Destroy()`, `v_i.Destroy()` in the example).
An example in the MFEM code for this is shown for BlockVector in https://github.com/mfem/mfem/pull/1468/commits/33b413042a132a03efb6412a6373d33eaa408a39.

However, we also concluded that another (long-term) solution should be implemented, as well: The dangling pointers from the aliases should behave as usual dangling pointers do, i.e. not make any trouble in the memory manager as long as they are not dereferenced. For that reason, I'll keep this issue open for now until we have some update.
